### PR TITLE
fix: add compatibility logic for empty array when unmarshal json string to route

### PR DESCRIPTION
**Pull request by okaybase**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues

___
### Bugfix
- Description
unmarshal json string to route error when route info contains "vars":{}. 

- How to fix?
add compatibility logic for empty array when unmarshal json string to route

___
### New feature or improvement
- Describe the details and related test reports.

___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

**Review by gxthrj — APPROVED** [review]

**Comment by tokers** [comment]
@okaybase Now you can re-factor this PR.

**Comment by okaybase** [comment]
> 
> 
> @okaybase Now you can re-factor this PR.

@tokers i have refactor this PR base on the #373

File: pkg/apisix/resource.go [hunk 1]
```diff
@@ -72,6 +72,27 @@ type item struct {
 	Value json.RawMessage `json:"value"`
 }
 
+type routeWrap v1.Route
```
Comment (tokers) [inline pkg/apisix/resource.go:75]: You can just create a type `Vars` in `pkg/types/apisix/v1`, and Implement the `json.Unmarshaler` interface for it.

Comment (okaybase) [reply pkg/apisix/resource.go]: Yes, thanks for your suggestion! I have rewrite.

File: pkg/apisix/resource_test.go [hunk 2]
```diff
@@ -74,3 +75,18 @@ func TestItemConvertRoute(t *testing.T) {
...
+	err := json.Unmarshal([]byte(data), &route)
+	assert.Nil(t, err)
+
+	data = `{"vars":{"a":"b"}}`
+	err = json.Unmarshal([]byte(data), &route)
+	assert.Equal(t, err.Error(), "unexpected non-empty object")
+
+	data = `{"vars":[]}`
+	err = json.Unmarshal([]byte(data), &route)
+	assert.Nil(t, err)
```
Comment (tokers) [inline pkg/apisix/resource_test.go:91]: Also add a normal case where `vars` is a non-empty slice.

File: pkg/types/apisix/v1/types.go [hunk 3]
```diff
@@ -79,16 +79,37 @@ type Metadata struct {
...
+	Uri         string   `json:"uri,omitempty" yaml:"uri,omitempty"`
+	Priority    int      `json:"priority,omitempty" yaml:"priority,omitempty"`
+	Vars        Vars     `json:"vars,omitempty" yaml:"vars,omitempty"`
+	Uris        []string `json:"uris,omitempty" yaml:"uris,omitempty"`
+	Methods     []string `json:"methods,omitempty" yaml:"methods,omitempty"`
+	RemoteAddrs []string `json:"remote_addrs,omitempty" yaml:"remote_addrs,omitempty"`
+	UpstreamId  string   `json:"upstream_id,omitempty" yaml:"upstream_id,omitempty"`
+	Plugins     Plugins  `json:"plugins,omitempty" yaml:"plugins,omitempty"`
+}
+
```
Comment (tokers) [inline pkg/types/apisix/v1/types.go:93]: Add a comment for it.

```suggestion
// Vars represents the route match expressions of APISIX.
type Vars [][]StringOrSlice
```

File: pkg/apisix/resource_test.go (hunk 2 above)
Comment (okaybase) [reply pkg/apisix/resource_test.go:91]: The test case has been added!

File: pkg/types/apisix/v1/types.go (hunk 3 above)
Comment (okaybase) [reply pkg/types/apisix/v1/types.go:93]: The comment for Vars type declare has been added.

**Review by tokers — APPROVED** [review]

**Comment by tokers** [comment]
@okaybase Merged, many thanks!

**Pull request merged**